typing/rec_check.ml

```diff
@@ -577,15 +577,30 @@ let rec expression : Typedtree.expression -> term_judg =
       *)
       expression arg << Guard
     | Texp_apply (e, args)  ->
+        (* [args] may contain omitted arguments, corresponding to labels in
+           the function's type that were not passed in the actual application.
+           The arguments before the first omitted argument are passed to the
+           function immediately, so they are dereferenced. The arguments after
+           the first omitted one are stored in a closure, so guarded.
+           The function itself is called immediately (dereferenced) if there
+           is at least one argument before the first omitted one, stored in
+           the closure (guarded) otherwise. *)
+        let rec split_args before_omitted = function
+          | (_, None) :: rest -> before_omitted, rest
+          | ((_, Some _) as arg) :: rest ->
+            split_args (arg :: before_omitted) rest
+          | [] -> before_omitted, []
+        in
+        let dereferenced, guarded = split_args [] args in
         let arg (_, eo) = option expression eo in
-        let app_mode = if List.exists is_abstracted_arg args
```

Comment:
`is_abstracted_args` is local to this file and used in another place that also looks wrong. Please have a look there and maybe also fix it -- the helper function can probably be removed then. (I think that that other place should now check whether the first argument is applied or not.)

Reply:
The other place (in the classification code) is fine, actually. As long as there is one omitted argument the whole expression will have a closure allocation in tail position.